Support index expressions and selectors in range loops

This won't work with the compiler we have:
```
ints := []int{5, 4, 3, 2, 1}
for ints[0] = range ints {
        break
}
return ints[0]
```
Right now that'd be
```
panic: interface conversion: ast.Expr is *ast.IndexExpr, not *ast.Ident [recovered]
        panic: interface conversion: ast.Expr is *ast.IndexExpr, not *ast.Ident
```
It'll be a bit more gentle error soon, but properly fixing this requires some refactoring of `*ast.AssignStmt` handler and that's not what I'd like to do right now for this very specific case.

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -806,26 +806,10 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			} else {
 				c.saveSequencePoint(n.Lhs[i].Pos(), n.Lhs[i].End())
 			}
-			switch t := n.Lhs[i].(type) {
-			case *ast.Ident:
-				if n.Tok == token.DEFINE {
-					if !multiRet {
-						c.registerDebugVariable(t.Name, n.Rhs[i])
-					}
-					if t.Name != "_" {
-						c.scope.newLocal(t.Name)
-					}
-				}
-				c.emitStoreVar("", t.Name)
-
-			case *ast.SelectorExpr:
-				c.emitStoreSelectorExpr(t) // store the field
-
-			// Assignments to index expressions.
-			// slice[0] = 10
-			case *ast.IndexExpr:
-				c.emitStoreIndexExpr(t)
+			if t, ok := n.Lhs[i].(*ast.Ident); ok && n.Tok == token.DEFINE && !multiRet {
+				c.registerDebugVariable(t.Name, n.Rhs[i])
 			}
+			c.emitStoreExpr(n.Lhs[i], n.Tok)
 		}
 		return nil
 
@@ -1500,25 +1484,21 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 		var (
 			haveKey   bool
 			haveVal   bool
-			keyIdent  *ast.Ident
+			keyExpr   = n.Key
 			keyLoaded bool
-			valIdent  *ast.Ident
+			valExpr   = n.Value
 		)
-		if n.Key != nil {
-			keyIdent, haveKey = n.Key.(*ast.Ident)
-			if !haveKey {
-				c.prog.Err = errors.New("only simple identifiers can be used for range loop keys (see #2870)")
+		if keyExpr != nil {
+			haveKey = c.validateRangeTarget(keyExpr)
+			if c.prog.Err != nil {
 				return nil
 			}
-			haveKey = (keyIdent.Name != "_")
 		}
-		if n.Value != nil {
-			valIdent, haveVal = n.Value.(*ast.Ident)
-			if !haveVal {
-				c.prog.Err = errors.New("only simple identifiers can be used for range loop values (see #2870)")
+		if valExpr != nil {
+			haveVal = c.validateRangeTarget(valExpr)
+			if c.prog.Err != nil {
 				return nil
 			}
-			haveVal = (valIdent.Name != "_")
 		}
 		if haveKey {
 			if isMap {
@@ -1530,10 +1510,7 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			} else {
 				emit.Opcodes(c.prog.BinWriter, opcode.DUP)
 			}
-			if n.Tok == token.DEFINE {
-				c.scope.newLocal(keyIdent.Name)
-			}
-			c.emitStoreVar("", keyIdent.Name)
+			c.emitStoreExpr(keyExpr, n.Tok)
 		}
 		if haveVal {
 			if !isMap || !keyLoaded {
@@ -1547,10 +1524,7 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 					opcode.SWAP, // key should be on top
 					opcode.PICKITEM)
 			}
-			if n.Tok == token.DEFINE {
-				c.scope.newLocal(valIdent.Name)
-			}
-			c.emitStoreVar("", valIdent.Name)
+			c.emitStoreExpr(valExpr, n.Tok)
 		}
 
 		ast.Walk(c, n.Body)
@@ -2628,6 +2602,43 @@ func (c *codegen) resolveFuncDecls(f *ast.File, pkg *types.Package) {
 			fs := c.newFunc(n)
 			fs.file = f
 		}
+	}
+}
+
+// validateRangeTarget checks that expr can be used as a range assignment target.
+// It returns false for the blank identifier.
+func (c *codegen) validateRangeTarget(expr ast.Expr) bool {
+	switch t := expr.(type) {
+	case *ast.Ident:
+		if t.Name == "_" {
+			return false
+		}
+		return true
+	case *ast.IndexExpr, *ast.SelectorExpr:
+		return true
+	default:
+		c.prog.Err = fmt.Errorf("unsupported range target %T, expected identifier, selector or index expression", expr)
+		return false
+	}
+}
+
+// emitStoreExpr stores the value on top of the stack into expr.
+func (c *codegen) emitStoreExpr(expr ast.Expr, tok token.Token) {
+	switch t := expr.(type) {
+	case *ast.Ident:
+		if tok == token.DEFINE && t.Name != "_" {
+			c.scope.newLocal(t.Name)
+		}
+		c.emitStoreVar("", t.Name)
+
+	case *ast.SelectorExpr:
+		c.emitStoreSelectorExpr(t)
+
+	case *ast.IndexExpr:
+		c.emitStoreIndexExpr(t)
+
+	default:
+		c.prog.Err = fmt.Errorf("unsupported assignment target %T", expr)
 	}
 }
 

--- a/pkg/compiler/for_test.go
+++ b/pkg/compiler/for_test.go
@@ -797,6 +797,128 @@ var forLoopTestCases = []testCase{
 		`,
 		big.NewInt(10),
 	},
+	{
+		"range, key to index expr",
+		`func F%d() int {
+			ints := []int{5, 4, 3, 2, 1}
+			for ints[0] = range ints {
+					break
+			}
+			return ints[0]
+		}
+		`,
+		big.NewInt(0),
+	},
+	{
+		"range, value to index expr",
+		`func F%d() int {
+			ints := []int{5, 4, 3}
+			dst := []int{0}
+			for _, dst[0] = range ints {
+				break
+			}
+			return dst[0]
+		}
+    	`,
+		big.NewInt(5),
+	},
+	{
+		"range, key and value to index expr",
+		`func F%d() int {
+			ints := []int{5, 4, 3}
+			dst := []int{100, 100}
+			for dst[0], dst[1] = range ints {
+				break
+			}
+			return dst[0] + dst[1]
+		}
+		`,
+		big.NewInt(5),
+	},
+	{
+		"range, key to selector",
+		`type S struct {
+			x int
+		}
+	
+		func F%d() int {
+			ints := []int{5, 4, 3}
+			s := S{x: 42}
+			for s.x = range ints {
+				break
+			}
+			return s.x
+		}
+		`,
+		big.NewInt(0),
+	},
+	{
+		"range, value to selector",
+		`func F%d() int {
+			ints := []int{5, 4, 3}
+			s := S{}
+			for _, s.x = range ints {
+				break
+			}
+			return s.x
+		}
+		`,
+		big.NewInt(5),
+	},
+	{
+		"range, key and value to selector",
+		`type KV struct {
+			k int
+			v int
+		}
+	
+		func F%d() int {
+			ints := []int{5, 4, 3}
+			kv := KV{}
+			for kv.k, kv.v = range ints {
+				break
+			}
+			return kv.k + kv.v
+		}
+		`,
+		big.NewInt(5),
+	},
+	{
+		"range, value index expr after full iteration",
+		`func F%d() int {
+			ints := []int{5, 4, 3}
+			dst := []int{0}
+			for _, dst[0] = range ints {
+			}
+			return dst[0]
+		}
+		`,
+		big.NewInt(3),
+	},
+	{
+		"range, key index expr after full iteration",
+		`func F%d() int {
+			ints := []int{5, 4, 3}
+			dst := []int{0}
+			for dst[0] = range ints {
+			}
+			return dst[0]
+		}
+		`,
+		big.NewInt(2),
+	},
+	{
+		"range, map key/value to index expr",
+		`func F%d() int {
+ 			m := map[int]int{7: 9}
+ 			dst := []int{0, 0}
+ 			for dst[0], dst[1] = range m {
+ 			}
+ 			return dst[0]*100 + dst[1]
+ 		}
+ 		`,
+		big.NewInt(709),
+	},
 }
 
 func TestForLoop(t *testing.T) {

